Runtime/Scripts/MeshSyncInstanceRenderer.cs

```diff
@@ -0,0 +1,151 @@
+        {
+            
+            var instances = data.FindUserProperty("instances");
+
+            if (instances.self == IntPtr.Zero)
+                return;
+
+            if (!obj.TryGetComponent(out MeshFilter meshFilter))
+            {
+                Debug.LogWarningFormat("Object {0} has instances but no MeshFilter", obj.name);
+                return;
+            }
+
+            if (!obj.TryGetComponent(out MeshRenderer renderer))
+            {
+                Debug.LogWarningFormat("Object {0} has instance but no MeshRenderer", obj.name);
+            }
+            
+            var mesh = meshFilter.sharedMesh;
+
+            var entry = meshInstances.Find(x => x.Mesh == mesh);
+            
+            if (entry == null)
+            {
+                entry = new MeshInstanceInfo
+                {
+                    Mesh = mesh
+                };
+                meshInstances.Add(entry);
+            }
+
+           
+
+            entry.Instances = DivideArrays(instances.matrixArray);
+            entry.Materials = renderer.sharedMaterials;
+            foreach (var mat in entry.Materials)
+            {
+                mat.enableInstancing = true;
+            }            
+        }
+
+        private List<Matrix4x4[]> DivideArrays(Matrix4x4[] arrays)
+        {
+            var result = new List<Matrix4x4[]>();
+            var iterations = arrays.Length / 1023;
+            for (var i = 0; i < iterations; i++)
+            {
+                var array = new Matrix4x4[1023];
+                
+                Array.Copy(
+                    arrays, 
+                    i * 1023, 
+                    array, 
+                    0, 
+                    1023);
+                
+                result.Add(array);
+            }
+
+            var remainder = arrays.Length % 1023;
+            if (remainder > 0)
+            {
+                var array = new Matrix4x4[remainder];
+                
+                Array.Copy(
+                    arrays, 
+                    iterations*1023, 
+                    array, 
+                    0, 
+                    remainder);
+                
+                result.Add(array);
+            }
+
+            return result;
+        }
+
+        private void Update()
+        {
+            foreach (var entry in meshInstances)
+            {
+                RenderInstances(entry);
+            }
+        }
+
+        private void RenderInstances(MeshInstanceInfo entry)
+        {
+            var mesh = entry.Mesh;
+            var matrixBatches = entry.Instances;
+
+            if (entry.Materials.Length == 0)
+                return;
+            
+            for (var i = 0; i < mesh.subMeshCount; i++)
+            {
+                // Try to get the material in the same index position as the mesh
+                // or the last material.
+                
+                var materialIndex = Mathf.Max(Mathf.Min(entry.Materials.Length -1, i), 0);
```

Comment:
Isn't this the same as `Mathf.Clamp` ?

Reply:
It is indeed, I changed it use the clamping function now